qiskit/circuit/quantumcircuit.py

```diff
@@ -2632,14 +2632,26 @@ def parameters(self) -> ParameterView:
     @property
     def num_parameters(self) -> int:
         """The number of parameter objects in the circuit."""
+        # Avoid a (potential) object creation if we can.
+        if self._parameters is not None:
+            return len(self._parameters)
         return len(self._unsorted_parameters())
```

Comment:
In the current format calling `num_parameters` would call `_unsorted_parameters` every time if `_parameters` is not yet set. Since sorting the parameters wouldn't be a big overhead once we gather them, how about replacing this whole property simply by
```
return len(self.parameters)
```
?

That would be a bit more expensive the first time, since it sorts the parameters on top of gathering them, but then the cached `_parameters` would be set.

Reply:
> The QuantumCircuit.parameters getter also (unfortunately) creates a new set object on every call, which is quite unfortunate for a property, really. It's potentially worth a follow-up PR to change the internal cache to a frozenset so we can return it directly, although it's annoying that that would be a breaking API change.

That would mean changing from a sorted `ParameterView` to returning a set of the parameters? Yeah would be worth a discussion whether sorting only when we really need to is better. A first issue that comes to mind is that users can no longer easily check the order of parameters and that binding via `dict(zip(circuit.parameters, values))` isn't valid anymore.

Reply:
I'm not proposing removing the sort from `QuantumCircuit.parameters`, this is just avoiding producing new temporary collections when we can get use data from something else already.  The other part is that the `parameters` field is marked as an attribute, but it actually does a non-trivial amount of work to get itself; even if `_parameters` is cached, it returns a copy of it.  It does that so a user can't mutate the inners of `QuantumCircuit`, but it comes with the cost of these performance bugs; it's not at all clear from accessing `qc.parameters` that what you're actually getting is amortised `qc._parameters.copy()`, which is somewhat costly for large numbers of parameters.